software/spmd/pc_profiler/profiler.S

```diff
@@ -0,0 +1,231 @@
+/* 
+  Description:
+  A software-based PC profiler which builds a histogram of PCs encountered by the program
+
+  Memory layout:
+  DRAM --> 16MB (configurable) for maintaing a histogram for N tiles. 
+
+  Registers used:
+  The program being profiled is free to use any registers within if there are no other imposed constraints.
+  The profiler uses the following registers - 
+  t0 --> Base address of the histogram array (actual value stored in dmem[1])
+  t1 --> Scratchpad (actual value stored in dmem[0])
+*/
+
+#include "bsg_manycore_arch.h"
+#include "bsg_manycore_asm.h"
+
+// Useful to control the amount of space required to simulate the use of the profiler.
+// We want 16MB when running on actual hardware.
+#ifndef HIST_SPACE
+#define HIST_SPACE 16777216
+#endif
+
+#ifndef PROFILE
+#define PROFILE 0
+#endif
+
+#if PROFILE > 0
+/* 
+  Data section:
+  Defines a 2 element (8 byte) array to store the base pointer of the histogram array and for use by the interrupt handler to spill some registers
+  dmem[0] = histogram base pointer, temporary storage
+  dmem[1] = temporary storage
+*/
+.section .dmem.interrupt, "aw"
+  _interrupt_arr: .space 8, 0x0
+
+/*
+  DRAM section:
+  Defines a 16MB (array size is configurable) chunk in DRAM to store the histogram.
+
+  Lingering question(s):
+  1. Assigned space per PC (i.e bin capacity) as 32-bits. Is this enough, too much or too less?
+  2. Base pointer in DRAM (for now it is mapped to start of DRAM) [Consider static assignment in linker]
+*/
+.section .dram, "aw"
+  _histogram_arr: .space HIST_SPACE, 0x0
+#endif
+
+/*
+  Startup:
+  This is very similar to the code in software/spmd/common/crt.S and adds only the interrupt service routine (ISR) and sets up the DMEM with the histogram base address. 
+*/
+.section .crtbegin
+.globl _start
+
+#if PROFILE > 0
+/*
+  Remote interrupt handler (should map to address 0)
+*/
+_remote_interrupt:
+  j _remote_interrupt_handler
+
+/*
+  Trace interrupt handler (should map to address 4)
+  Clears the pending trace interrupt bit in the MIP CSR and atomically increments the histogram bin count based on the MEPC CSR value
+
+  The bin index in the histogram can be found using the following equation index = base ptr + EPC
+  (Although EPC is 4-byte aligned, we are assigning 4 bytes per bin, so we don't need to do EPC >> 2)
+
+  Lingering concerns:
+  Since we are using the MEPC, we don't have the PC of the first instruction in the program. 
+  (Probably does not matter because it's just 1 datapoint)
+*/
+_trace_interrupt:
```

Comment:
So are both address 0x4 and 0x0 special addresses now? So specific interrupt handlers live at special code addresses?

Reply:
Yup. They should be as per the interrupt specification that was given. (Not able to share the doc here since I am not a owner, but if you have access to a common drive where all the HB docs are present, look for the file - HammerBlade Machine Interrupts. That should give all the interrupt specific info)

Reply:
https://docs.google.com/document/d/1HKa3Qts55TLqltLKpoAalcVJokGHOQm27u-1FXmbTQE/edit?ts=60075876

@mrutt92 high-level: when you take a trap, you go to BASE + 4*CAUSE. We hardcode BASE to -4*TRACE_CAUSE so that the handler pointers get placed at the bottom of DMEM

Reply:
So are there only two causes? Shouldn't this `_trace_interrupt` be one instruction to jump to another location like with `_remote_interrupt`? Or is this only meant for a specific test?

Reply:
@dpetrisko am I misunderstanding something? Do you mean DMEM or IMEM in this case? I see that we've also hardcoded DMEM addresses for interrupt handler storage, but I think you're talking about PC addresses.

Reply:
Disregard, I misread your question